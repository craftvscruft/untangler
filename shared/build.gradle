plugins {
    id 'java'
}
repositories {
    mavenCentral()
}
sourceCompatibility = JavaVersion.VERSION_16
targetCompatibility = JavaVersion.VERSION_16
dependencies {
    implementation "org.apache.commons:commons-lang3:${property('commons-lang3.version')}"
    implementation "com.fasterxml.jackson.core:jackson-annotations:${property('jackson.version')}"
}