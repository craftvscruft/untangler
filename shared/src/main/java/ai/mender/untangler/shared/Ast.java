package ai.mender.untangler.shared;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Stream;

@JsonInclude(JsonInclude.Include.NON_EMPTY)
public record Ast(String kind, SourceRange range, SourceRange nameRange, List<Ast> children, Map<String, String> props) {

    public static Ast create(String kind, SourceRange range, SourceRange nameRange) {
        return new Ast(kind, range, nameRange, new ArrayList<>(), new HashMap<>());
    }

    @JsonIgnore
    public Stream<Ast> dfsStream() {
        return Stream.concat(Stream.of(this), children.stream().flatMap(Ast::dfsStream));
    }
}
