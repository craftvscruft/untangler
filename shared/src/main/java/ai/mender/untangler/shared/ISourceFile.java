package ai.mender.untangler.shared;

import java.io.Reader;
import java.nio.charset.Charset;

public interface ISourceFile {
    Reader getReader();

    String getName();


    Charset getCharset();
}
