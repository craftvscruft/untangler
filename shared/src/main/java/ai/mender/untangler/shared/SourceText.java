package ai.mender.untangler.shared;

public record SourceText(String text, SourceRange range) {
}
