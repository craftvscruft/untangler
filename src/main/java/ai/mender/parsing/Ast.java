package ai.mender.parsing;

import ai.mender.domain.SourceRange;
import com.google.common.base.Objects;

import java.util.List;
import java.util.Optional;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public record Ast(String tag, Ast parent, List<Ast> children, SourceRange range){

    public void addChild(Ast child) {
        children().add(child);
    }
    public boolean isChildless() {
        return children.isEmpty();
    }

    public boolean allChildrenAreChildless() {
        return children.stream().allMatch(Ast::isChildless);
    }

    public Stream<Ast> reverseChildrenStream() {
        var children = children();
        IntStream reverseIndices = IntStream.range(0, children.size()).map(i -> children.size() - i - 1);
        return reverseIndices.mapToObj(children::get);
    }

    public Optional<Ast> getFirstChildByTag(String tag) {
        return children().stream().filter((child) -> tag.equals(child.tag())).findFirst();
    }

    public List<String> childTags() {
        return children.stream().map(Ast::tag).toList();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Ast ast = (Ast) o;
        return Objects.equal(tag, ast.tag) && Objects.equal(children, ast.children) && Objects.equal(range, ast.range);
    }

    @Override
    public int hashCode() {
        // Do not include parents.
        return Objects.hashCode(tag, children, range);
    }

    @Override
    public String toString() {
        return "Ast{" +
                "tag='" + tag + '\'' +
                ", children=" + childTags() +
                ", range=" + range +
                '}';
    }
}
