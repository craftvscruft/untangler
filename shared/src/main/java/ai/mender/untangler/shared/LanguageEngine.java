package ai.mender.untangler.shared;

import ai.mender.untangler.shared.response.*;

import java.util.List;
import java.util.stream.Stream;

public interface LanguageEngine {
    Stream<FunctionRec> functions(ISourceFile sourceFile);
    Stream<CommentRec> comments(ISourceFile sourceFile);

    default SourceEditListResponse insertComment(int line, String text) {
        String commentText = formatMultiLineComment(text.trim());
        SourceEdit edit = SourceEdit.insert(new SourcePosition(line, 1), commentText);
        return new SourceEditListResponse(true, "OK", List.of(edit));
    }

    String formatMultiLineComment(String text);
}
