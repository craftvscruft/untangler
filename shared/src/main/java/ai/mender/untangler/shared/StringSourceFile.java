package ai.mender.untangler.shared;

import ai.mender.untangler.shared.ISourceFile;

import java.io.Reader;
import java.io.StringReader;
import java.nio.charset.Charset;

public class StringSourceFile implements ISourceFile {
    private String name;
    private String source;

    public StringSourceFile(String name, String source) {
        this.name = name;

        this.source = source;
    }


    @Override
    public Reader getReader() {
        return new StringReader(source);
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public Charset getCharset() {
        return Charset.defaultCharset();
    }

    public String getSource() {
        return source;
    }

    public void setSource(String source) {
        this.source = source;
    }
}
