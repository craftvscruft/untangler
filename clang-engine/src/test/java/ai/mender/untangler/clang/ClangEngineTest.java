package ai.mender.untangler.clang;

import ai.mender.untangler.shared.Ast;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.ObjectWriter;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.io.FileReader;
import java.io.IOException;

class ClangEngineTest {

    public static final String BUILD_RESOURCES_PATH = "clang-engine/build/resources/test";
    public static final String RESOURCES_PATH = "clang-engine/src/test/resources";

    private String getTestResourcePath(String fileName) {
        ClassLoader classLoader = getClass().getClassLoader();
        File file = new File(classLoader.getResource(fileName).getFile());
        return file.getAbsolutePath();
    }

    private static String convertBuildResourcePathToSourcePath(String path) {
        return path.replace(BUILD_RESOURCES_PATH, RESOURCES_PATH);
    }

    @Test
    public void filterOutsideFileFromClangAst() throws IOException {
        String clangAstPath = getTestResourcePath("clang-ast-system-header-example.json");
        ObjectMapper objectMapper = new ObjectMapper();
        JsonNode clangAst = objectMapper.reader().readTree(new FileReader(clangAstPath));
        JsonNode clangAstFiltered = ClangEngine.filterClangAstByFile(clangAst, new ClangParseConfig("/tmp/class.cpp"));
        Approvals.verify(objectMapper.writer().withDefaultPrettyPrinter().writeValueAsString(clangAstFiltered));
    }

    @Test
    public void simpleCFileToAst() throws IOException {
        approveAstTransformation("onevar.c");
    }

    @Test
    public void structCExampleToAst() throws IOException {
        approveAstTransformation("struct.c");
    }

    @Test
    public void classCppExampleToAst() throws IOException {
        approveAstTransformation("class.cpp");
    }

    private void approveAstTransformation(String fileName) throws IOException {
        // Generate and save clang ast if it's missing but otherwise use the one that's there,
        // so we can run the tests without clang.
        String codePath = getTestResourcePath(fileName);
        String clangAstSourcePath = convertBuildResourcePathToSourcePath(codePath + ".clang-ast.json");
        File clangAstSourchFile = new File(clangAstSourcePath);
        if (!clangAstSourchFile.exists()) {
            System.out.println("Writing " + clangAstSourcePath);
            ClangParseConfig config = new ClangParseConfig(codePath);
            JsonNode clangAst = ClangEngine.dumpClangAstJson(config);
            JsonNode filteredClangAst = ClangEngine.filterClangAstByFile(clangAst, config);
            ObjectWriter writer = new ObjectMapper().writer().withDefaultPrettyPrinter();
            writer.writeValue(clangAstSourchFile, filteredClangAst);
        }
        JsonNode clangAst = new ObjectMapper().reader().readTree(new FileReader(clangAstSourchFile));
        Ast ast = ClangEngine.transformClangAstToUnifiedAst(clangAst, new ClangParseConfig(codePath));
        Approvals.verify(Main.getAstJsonWriter().writeValueAsString(ast));
    }
}