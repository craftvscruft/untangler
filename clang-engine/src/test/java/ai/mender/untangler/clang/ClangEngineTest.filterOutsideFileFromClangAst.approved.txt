{
  "id" : "0x122841608",
  "kind" : "TranslationUnitDecl",
  "loc" : { },
  "range" : {
    "begin" : { },
    "end" : { }
  },
  "inner" : [ {
    "id" : "0x1302bb2b8",
    "kind" : "CXXRecordDecl",
    "loc" : {
      "offset" : 24,
      "file" : "/tmp/class.cpp",
      "line" : 2,
      "col" : 7,
      "tokLen" : 6
    },
    "range" : {
      "begin" : {
        "offset" : 18,
        "col" : 1,
        "tokLen" : 5
      },
      "end" : {
        "offset" : 179,
        "line" : 9,
        "col" : 1,
        "tokLen" : 1
      }
    },
    "isReferenced" : true,
    "name" : "widget",
    "tagUsed" : "class",
    "completeDefinition" : true,
    "definitionData" : {
      "canConstDefaultInit" : true,
      "copyAssign" : {
        "hasConstParam" : true,
        "implicitHasConstParam" : true,
        "needsOverloadResolution" : true,
        "trivial" : true
      },
      "copyCtor" : {
        "hasConstParam" : true,
        "implicitHasConstParam" : true,
        "needsOverloadResolution" : true,
        "trivial" : true
      },
      "defaultCtor" : {
        "defaultedIsConstexpr" : true
      },
      "dtor" : {
        "needsOverloadResolution" : true,
        "nonTrivial" : true,
        "simple" : true
      },
      "hasUserDeclaredConstructor" : true,
      "isStandardLayout" : true,
      "moveAssign" : {
        "exists" : true,
        "needsOverloadResolution" : true,
        "nonTrivial" : true,
        "simple" : true
      },
      "moveCtor" : {
        "exists" : true,
        "needsOverloadResolution" : true,
        "nonTrivial" : true,
        "simple" : true
      }
    },
    "inner" : [ ]
  } ]
}